docs/rfcs/generic-record-sync.md

```diff
@@ -0,0 +1,1287 @@
+   from the local record not present in "the delta"
+
+    - It doesn't matter if you take from local or remote here, since
+      these are the fields that we just determined were equal.
+
+2. For each composite field containing one or more non-deprecated subfields
+   present in "the delta": the merge is performed roughly the same as the 3WM
+   case.
+
+    - If the composite root has `UntypedMerge::PreferRemote`, then prefer
+      remote for all members of the composite.
+
+    - If the composite root has `UntypedMerge::TakeNewest`, then copy the fields
+      in the composite from whichever has been modified more recently (as
+      determined by modification timestamps) between the local and incoming
+      remote records into "the merged record".
+
+        - Note that we're copying from the records, not from the deltas.
+
+    - If the composite root has `NumberMerge::TakeMin` or `NumberMerge::TakeMax`
+      as it's merge strategy, then compare the values of the composite root in
+      the local and remote records numerically, and copy all fields in the
+      composite from the winner to "the merged record".
+
+3. For remaining (non-composite) fields present in "the delta",
+   "incoming delta", Store the result of the following in "the merged record":
+    - `UntypedMerge::PreferRemote`, `UntypedMerge::TakeNewest`, `UntypedMerge::Duplicate`, `NumberMerge::TakeMax`, `NumberMerge::TakeMin`, `BooleanMerge::PreferTrue`, `BooleanMerge::PreferFalse`, `OwnGuid`, `ForeignGuid`:
+        - All handled as in 3WM.
+    - `NumberMerge::TakeSum`: The larger value between local and remote is taken
+        - This preserves the fact that nobody ever sees the value decrease.
+    - `UntypedMap { prefer_deletions }`: The maps are merged, breaking ties in
+      favor of the more recently modified.
+        - if `prefer_deletions` is true, any field represented by a tombstone in
+          either side is a tombstone in the output
+    - `RecordSet { dedupe_key, prefer_deletions }`:
+        - The result is the set union of the two, with deletions preferred if
+          `prefer_deletion` is true
+
+4. The "merged delta" is applied to the mirror record to produce the new mirror
+   record which will be uploaded to the server.
+
+    - This record will have a vector clock that is a descendent of the local,
+      mirror, and incoming remote clocks.
+
+## New metadata records:
+
+These are not stored in meta/blah, they're stored in
+`$collection/__metadata__:blah` (or, `$collection/__metadata__%3Ablah`, really).
+This is to allow transactional behavior during updates, which is important for
+the case of schema changes. Existing clients will need to change to ignore these
+records.
+
+(I'm willing to bikeshed over the name so long as it doesnt fit any pattern any
+current client generates)
+
+**Important**: This requires a change to current clients so that they know to ignore all items
+in a collection whose ID starts with `__metadata__:`!
```

Comment:
I'd love to do this, but the timing seems tricky. It feels like the first version of this is likely be ready before the rust server is widely usable, so new server features are tricky.

Also, we're no longer a unified sync server / client team :(

This part of the design has gotten a little worse, if I'm being honest, since I've moved it outside of this collection, which means transactional updates of the schema and records will not be  possible, unfortunately.